cachi2/interface/cli.py

```diff
@@ -405,6 +405,94 @@ def all_files_are_jsons(sbom_files_to_merge: Paths) -> Paths:
     return all_files_are_jsons(enough_files_for_merge(list(set(sbom_files_to_merge))))
 
 
+def merge_relationships(
+    relationships_list: List[List[SPDXRelation]], doc_ids: List[str], packages: List[SPDXPackage]
```

Comment:
A list is a sequence of variable length, the docstring states that the function takes two relationship lists. If it is just two then it  would make more sense to have 'this' and 'that' r_lists and to merge them. Or some better name which would let everyone see that there are two distinct entities to be processed. 'main' and 'other' are good candidates given that those are already used in the code.

If it is not about merging two lists then I strongly recommend making it about merging two lists and then reducing any list of lists with it if needed.

Reply:
So let me put it this way. Use case you need to implement is to be able to merge multiple files. I as developer I see more convenient to have method which does that instead of having method which merge two files and then I need to think how to actually use it for multiple files. Will it produce same output as if there would be method for multiple files? I cannot be sure until I review body of the method because there could be something which affects the output.
And for the efficiency, approach of merging multiple lists in one call is just more efficient then calling same method multiple times. Regardless of if it's going to be used often or not.
Implementation of merging multiple lists in one method is very similar to how it would be implemented for two lists. So overall there are more positives than negatives.

Reply:
Yes, you are right. As a developer of a library method you can (and arguably should) provide a method to merge multiple SBOMs together. Implementation of that method in terms of a merger for two SBOms is trivial and I have posted a (slightly incorrect) version above: you simply reduce a two-merger with a list of SBOMs and None (not with a list as I have incorrectly stated before). Now the two-merger should accept two optional SBOMs and return an optional SBOM as a result. This gives us three base cases 1) merge two Nones -- should result in None (a degenerate case, but must be accounted for if we are serious about our types); 2) merge a SBOM with None -- just return the SBOM as is and lastly 3) merge two actual SBOMs. Reduction will always product correct results if two-merger produces correct results.

Appeals to efficiency must be supported by data. Inefficient methods that get called once a day, take a full second to complete in the worst case, but take three minutes to figure out when there is a bug are fundamentally better than super-efficient methods that run for just 50 milliseconds every day, but take eight hours to comprehend. Two-merger might be easier to reason about: it will likely have simple `if` rules which are easier to follow than loops or LCs. Here I suggest trading some efficiency for ease of maintenance: greater readability usually increases some multiplicative factor of running time while greater efficiency often results in much harder maintenance. All efficiency aspects must be thoroughly commented and explained, or it will become an unmaintainable mess really fast.